x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/conditional/Case.java

```diff
@@ -185,6 +184,29 @@ public Object fold() {
         return elseValue.fold();
     }
 
+    /**
+     * Fold the arms of {@code CASE} statements.
+     * <pre>{@code
+     * EVAL c=CASE(true, foo, bar)
+     * }</pre>
+     * becomes
+     * <pre>{@code
+     * EVAL c=foo
+     * }</pre>
+     */
+    public Expression partiallyFold() {
+        for (Condition condition : conditions) {
+            if (condition.condition.foldable() == false) {
+                return this;
+            }
+            Boolean b = (Boolean) condition.condition.fold();
+            if (b != null && b) {
+                return condition.value;
+            }
+        }
+        return elseValue;
+    }
+
```

Comment:
I don't think this is a normal way to do rules but it feels like a nice way to test things.

Reply:
I find it easier to put this logic inside the rule and use `replaceChildren` to create a new instance (see `SimplifyCase`) but this one works as well.

Reply:
I wanted to try it and see if it made it easier to test and read. I *like* that you can test it in unit tests for `CASE`. I still made a little one in the optimizer tests to make sure it's plugged in though.

I don't like that it's different from everything else, but maybe it's a useful pattern?